File: 88-merge-sorted-array/88-merge-sorted-array.py
import copy
class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        """
        Do not return anything, modify nums1 in-place instead.
        Easy Appraoch: Just put them together and sort nums1
        """
        # for i in range(n):
        #     nums1[m+i] = nums2[i]
        # nums1.sort()
        
        """
        Using 3 pointer along with extra space
        """
        copy_nums1 = copy.deepcopy(nums1[:m])
        
        #read pointers
        read_copy = 0
        read_nums2 = 0      
        for i in range(m+n):
            # First Check: If pointer for nums2 is out of bound auto into first
            # Or
            # Second Check: Nums1copy ptr is not out of bounds and the value is smaller than in the 
            # nums2
            if (read_nums2 >=n) or (read_copy < m and copy_nums1[read_copy] < nums2[read_nums2]):
                nums1[i] = copy_nums1[read_copy]
                read_copy+=1
            else:
                nums1[i] = nums2[read_nums2]
                read_nums2+=1
        
        
        
        

File: 88-merge-sorted-array/88-merge-sorted-array.cpp
class Solution {
public:
    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        
        vector<int> result(m+n);
        int c1 = 0;
        int c2 = 0;
        for(int i=0; i<m+n; i++)
        {
            if(c1>=m)
            {
                result[i] = nums2[c2++];
                continue;
            }
            if(c2>=n)
            {
                result[i] = nums1[c1++];
                continue;
            }
            
            int tmp = -1;
            
            if(nums1[c1] < nums2[c2])
            {
                tmp = nums1[c1];
                c1++;
            }
            else
            {
                tmp = nums2[c2];
                c2++;
            }
            
            result[i] = tmp;
        }
        nums1 = result;
    }
};